### ISSUE TITLE
Fl_Window::default_icon(Fl_RGB_Image *image) produces garbage icon when image is scaled

### ISSUE THREAD

**Issue by wcout** [fltk/fltk#205]
I have run incidentally into the issue, using a previously scaled image (by scale() method) as window default icon.
The icon (seen in the taskbar) does then contain garbage (Ubuntu 20.04). There seems also to be potential to crash the application.

Here's a test program. It assumes to be run from the test folder and is using` images/tiny.png`.

````cxx
#include <FL/Fl_Window.H>
#include <FL/Fl_PNG_Image.H>
#include <FL/Fl.H>
#include <cassert>

void setIcon( Fl_RGB_Image *image_ )
{
//	image_->normalize(); // has no effect on this issue
	Fl_Window::default_icon( image_ );
}

int main()
{
	Fl_Window win(100, 100);
	Fl_PNG_Image icon_image("images/tiny.png");
	assert(icon_image.w() > 0);

	// this makes the subsequent setIcon() produce a garbage icon
	icon_image.scale(128, 128, 1, 1);

	setIcon( &icon_image );
	win.show();
	return Fl::run();
}
````

**Comment by Albrecht-S** [comment]
Thanks for the report. Confirmed (is a bug), found the culprit, and working on a fix ...

**Comment by Albrecht-S** [comment]
See attached patch for a working solution: [issue_205_diff_v1.txt](https://github.com/fltk/fltk/files/6191950/issue_205_diff_v1.txt). It's a proof of concept, you may try it, but I'm not satisfied with the code as-is. Way too many casts. Looking into simplifying (and clarifying) it...

**Comment by Albrecht-S** [comment]
FTR: copying rather than scaling the image to the requested size worked. A tiny modification to the demo program above solved the issue ... but currently a new commit (bfae813c4e15ec0a45a7b2cdecf32c3aac2776eb) introduced a regression that this wouldn't work anymore. Anyway, following is my (previously) working patch to the demo program as a proof of concept (this should be done internally in the FLTK library though).
```c++
void setIcon(Fl_RGB_Image *image)
{
#if (0)	// original test code
	// image->normalize(); // has no effect on this issue
	Fl_Window::default_icon(image);
#else	// modified:
	// ... but image->copy(W, H) has !
	image = (Fl_RGB_Image *)image->copy(image->w(), image->h());
	Fl_Window::default_icon(image);
	image->release(); // release copied image
#endif
}
```

**Comment by ManoloFLTK** [comment]
Attached is a proposed patch to fix this issue.
The bad effects of commit bfae813 have already been removed.
The patch completes what is still needed, i.e., support for default window icons.
It also contains some modifications of test/device.cxx to test
- per window and default icons
- scaled or unscaled images
- RGB or SVG images

All these combinations have been tested on the 4 platforms (although window icons have no Wayland implementation at this point).

[scaled-icon.patch.txt](https://github.com/fltk/fltk/files/8481743/scaled-icon.patch.txt)

**Comment by Albrecht-S** [comment]
@ManoloFLTK Thanks for the patch. Please don't commit/push yet, I need to take a look at it later tonight.

My intention was to make a `copy()` of the image **in cross-platform code** to avoid similar functionality in platform specific code. I didn't have the time to look further into your patch but I got the impression that it does the latter.

I've also been working on a patch to make `Fl_Image::copy()` 'const' which is now committed in 10537b7143a06a4c0c1c4caba4f0150f9e290090. This is essential to copy read-only ('const') images as those given to `Fl_Window::default_icon()` and similar API's like setting the cursor (which I'm also working on and I have a working prototype).

I'll post my patch later.

**Comment by ManoloFLTK** [comment]
@Albrecht-S : Yes, my patch does similar copies twice, in X11 and in Win32 code. Thing is that no copy is used for macOS.

**Comment by wcout** [comment]
> I've also been working on a patch to make Fl_Image::copy() 'const' which is now committed in https://github.com/fltk/fltk/commit/10537b7143a06a4c0c1c4caba4f0150f9e290090. This is essential to copy read-only ('const') images as those given to Fl_Window::default_icon() and similar API's like setting the cursor (which I'm also working on and I have a working prototype).

I don't think this is a good move. Why should an abstract copy(W,H) implementation not be allowed to modify it's variables? This change affects at least my `Fl_Animated_GIF_Image` implementation, which does not compile with it any longer...

**Comment by Albrecht-S** [comment]
> Why should an abstract copy(W,H) implementation not be allowed to modify it's variables? This change affects at least my Fl_Animated_GIF_Image implementation, which does not compile with it any longer...

@wcout Quick answer: because a copy operation should not modify the source operand. If this is really an issue and you can't change your `Fl_Animated_GIF_Image` then please open another issue for this. I'd be interested to see what breaks and why it breaks. Maybe you need to declare your `Fl_Animated_GIF_Image::copy()` methods `'const'` as well as [documented in the latest docs](https://fltk.gitlab.io/fltk/migration_1_4.html#migration_1_4_copy_image). I was aware of this issue but I believe this is a small price we need to pay. I'm open for discussion though, but please in another issue or - even better - in fltk.coredev. Thank you for your comment, this is always appreciated.

**Comment by Albrecht-S** [comment]
@ManoloFLTK I didn't forget your patch but it's too late now to check it today, I'll try tomorrow. One question though, you wrote:
> Thing is that no copy is used for macOS.

What do we gain with this additional maintenance burden - having three different implementations rather than one copy operation? I believe that performance or memory usage are not an issue here because setting an icon would be used rarely. Do you expect better icon graphics/resolution, or whatever? Thanks for your insights.

**Comment by Albrecht-S** [comment]
@wcout PS: I just rebased your branch on current master (1e99a00c65d3eaba724a3743dd79c25996756521) w/o merge conflicts and built the rebased branch on Linux successfully w/o compilation errors. However, I didn't test anything and - as I wrote above - the semantics may have changed because the virtual methods don't map correctly to the base class's methods anymore.

**Comment by ManoloFLTK** [comment]
@Albrecht-S 
> One question though, you wrote:
> 
>     Thing is that no copy is used for macOS.
> 
> What do we gain with this additional maintenance burden - having three different implementations rather than one copy operation? I believe that performance or memory usage are not an issue here because setting an icon would be used rarely. 

I agree.

> Do you expect better icon graphics/resolution, or whatever? 

I agree that's not an issue for as tiny a thing as a window icon.

I attach a diff file for changes for the alternative approach where as much work as possible is done in the platform-independent code. I'm not sure which approach is best and most maintainable and leave you choose and commit. (The changes in test/device.cxx therein are not intended for final commit, just to test setting window icons).

[scaled-iconsV2.patch.txt](https://github.com/fltk/fltk/files/8487491/scaled-iconsV2.patch.txt)
.

**Comment by MatthiasWM** [comment]
The patch works for me. I am not too worried about the few lines of code duplication, I am just hoping that we don't need to touch that code ever again ;-)

I put this on a "positive" timeout in the sense that I will apply the patch in two weeks, unless someone speaks up agains it ;-).

So **APPLY PATCH** on Dec 23rd it is.

**Comment by Albrecht-S** [comment]
@MatthiasWM  Sorry, I lost track of this. I'm no longer following (and won't test) this issue for lack of time, and I deassigned myself. However, there were two patches posted by @ManoloFLTK ...

Which patch would you like to apply? The first ([scaled-icon.patch.txt](https://github.com/fltk/fltk/files/8481743/scaled-icon.patch.txt)) in https://github.com/fltk/fltk/issues/205#issuecomment-1098051346 or the second ([scaled-iconsV2.patch.txt](https://github.com/fltk/fltk/files/8487491/scaled-iconsV2.patch.txt)) in https://github.com/fltk/fltk/issues/205#issuecomment-1098847254 ?

... and another one earlier by me, but mine was probably incomplete or incorrect (we may ignore that one).

**Comment by MatthiasWM** [comment]
Thanks for the reply and deassign. I will get to this in the next days,

**Comment by MatthiasWM** [comment]
Did not get around to fix this. Extending to Jan 10th
